minicluster/src/main/java/tachyon/master/AbstractLocalTachyonCluster.java

```diff
@@ -87,12 +93,100 @@ public void start(TachyonConf conf) throws IOException, ConnectionFailedExceptio
+    long startTime = System.currentTimeMillis();
+    String actionMessage = "waiting for master to serve web";
+    LOG.info(actionMessage + ELLIPSIS);
+    while (!isServing(getMaster().getWebBindHost(), getMaster().getWebLocalPort())) {
+      waitAndCheckTimeout(startTime, actionMessage);
+    }
+    actionMessage = "waiting for master to serve rpc";
+    LOG.info(actionMessage + ELLIPSIS);
+    while (!isServing(getMaster().getRPCBindHost(), getMaster().getRPCLocalPort())) {
+      waitAndCheckTimeout(startTime, actionMessage);
+    }
+  }
+
+  /**
+   * Waits for the worker to be ready.
+   *
+   * Specifically, waits for the worker to register with the master and for it to be possible to
+   * connect to the worker's data, rpc, and web ports.
+   */
+  private void waitForWorkerReady() {
+    long startTime = System.currentTimeMillis();
+    String actionMessage = "waiting for worker to register with master";
+    LOG.info(actionMessage + ELLIPSIS);
+    while (!workerRegistered()) {
+      waitAndCheckTimeout(startTime, actionMessage);
+    }
+    actionMessage = "waiting for worker to register with master";
+    LOG.info(actionMessage + ELLIPSIS);
+    while (!isServing(mWorker.getWebBindHost(), mWorker.getWebLocalPort())) {
+      waitAndCheckTimeout(startTime, actionMessage);
+    }
+    actionMessage = "waiting for worker to serve data";
+    LOG.info(actionMessage + ELLIPSIS);
+    while (!isServing(mWorker.getDataBindHost(), mWorker.getDataLocalPort())) {
+      waitAndCheckTimeout(startTime, actionMessage);
+    }
+    actionMessage = "waiting for worker to serve rpc";
+    LOG.info(actionMessage + ELLIPSIS);
+    while (!isServing(mWorker.getRPCBindHost(), mWorker.getRPCLocalPort())) {
+      waitAndCheckTimeout(startTime, actionMessage);
+    }
+  }
+
+  /**
+   * Checks whether the time since startTime has exceeded the maximum timeout, then sleeps for
+   * {@link CLUSTER_READY_POLL_INTERVAL_MS}ms
+   *
+   * @param startTime the time to compare against the current time to check for timeout
+   * @param actionMessage a message describing the action being waited for; this message is included
+   *        in the error message reported if timeout occurs
+   */
+  private void waitAndCheckTimeout(long startTime, String actionMessage) {
+    if (System.currentTimeMillis() - startTime > CLUSTER_READY_TIMEOUT_MS) {
+      throw new RuntimeException("Failed to start cluster. Timed out " + actionMessage);
+    }
+    CommonUtils.sleepMs(CLUSTER_READY_POLL_INTERVAL_MS);
+  }
+
+  /**
+   * @return whether the worker has registered with the master
+   */
+  private boolean workerRegistered() {
+    return WorkerIdRegistry.getWorkerId() != WorkerIdRegistry.INVALID_WORKER_ID;
+  }
+
+  /**
+   * @param host the host to try to connect to
+   * @param port the port to try to connect on
+   * @return whether a socket connection can be made to the given host on the given port
+   */
+  private static boolean isServing(String host, int port) {
```

Comment:
maybe as a TODO:

I feel this method belongs to network utils in common module.